### DISCUSSION TITLE
Allow same column to be selected multiple times in y-axis selections

### DISCUSSION THREAD

**Discussion by mhkeller** [perspective-dev/perspective#2577] [Q&A]
### Description of Problem:
Currently when doing aggregate functions, I can't compare different aggregate functions on the same column such as mean and median.

### Potential Solutions:
This would be remedied if I could add the same column to the `Y Axis` selection multiple times.

For example, using [this data](https://gist.github.com/mhkeller/8831f579f814af2a7415e043457d4e56) of student exams, I want to group by `exam_subject` and create a bar chart showing `avg` and `median` `exam_score` in each group. Once I select `exam_score` once, though, I can't select it again.

![Screenshot 2024-03-21 at 10 46 52](https://github.com/finos/perspective/assets/498744/b48391a3-d115-4ffc-9bcb-285a505f93d6)

**Comment by texodus** [comment]
You can duplicate a column in Perspective by giving it an alias:

```javascript
viewer.restore({
    columns: ["My Column", "My Column 2"],
    expressions: {
        "My Column 2": "\"My Column\""
    }
});
```

We're going to add quick-duplicate action buttons to the columns settings bar at some point, but they will just perform this action.

(Moving to discussions)

**Reply by mhkeller** [reply]
Got it. From a user perspective, some questions that come to mind are a) how to avoid having a bunch of differently named columns that are actually the same column and b) how to clean these up quickly once the user doesn't need them anymore. One UX flow could be:

1. User has avg `exam_score` in the Y-bar and they want median as well
2. User clicks a button or right-click's on `exam_score` box to duplicate it and the second `exam_score` auto-populates in the Y-bar area with the same name `exam_score` or a small HTML label that says it's a copy. It would be nice to avoid having `exam_score1` `exam_score2` etc. since that could be confusing. I could imagine coming to a project a couple weeks or months later and not remembering if those fields were the same or actually represented different data.
3. Once the user deselects the second `exam_score` column from the Y-bar field, that calculated column is deleted. The field was purpose-specific and the user doesn't have multiple copies of columns hanging around.

Out of curiosity, what's the limitation in having perspective do something under the hood where it allows the user to select `exam_score` again from the dropdown and internally, it creates an alias named however perspective's internals require but keeps the display field to the user as `exam_score`?

**Comment by texodus** [comment]
Holding the `Shift` key while clicking the reset button will delete all expression columns (this isn't an complete answer, it's just a tip). Expression columns actually have no performance/memory cost if they are not "in" the query, when they are deselected they are not calculated. The fact that in the UI they need an extra "step" to be deleted is just a convenience so you don't have to re-type the expression.

Perspective internally uses column name string itself as the unique column identifier, so we cannot physically have duplicate column names, even as expression aliases. The impact of this design decision trickles all the way down to e.g. the persistence and internal wire formats, and would be a major breaking change. 

Additionally, this is a _feature_ breaking change in that many Perspective-based systems are designed around the contract of a viewer config JSON that works when the row data changes. As most data formats (even Arrow) are not structured with aliases in mind, Perspective cannot deterministically rehydrate a Viewer config for a dataset if there are duplicate column names because it can't tell which columns go with which config aliases.

Like I said, we have some UX features in the works to make this a little quicker, and it is (I think) possible to make this design mostly transparent - we can duplicate the column implicitly when you drag/drop to a "duplicate" slot e.g., but this requires some careful consideration of the drag/drop rules.

**Reply by mhkeller** [reply]
Makes sense. I think something where it kept a unique name internally but the same name display name for the user would be a good solution from a user perspective. My mentioning of cleaning up was purely from a user perspective and not related to performance.

**Reply by timkpaine** [reply]
duplicate names are also useful when you want to have e.g. a data column and a spark bar of the same column
